### ISSUE TITLE
[RedisIdField] issue when using type different then string, GUID or Ulid 

### ISSUE THREAD

**Issue by Zlejsi** [redis/redis-om-dotnet#286]
When using different type then string, GUID or Ulid at [RedisIdField] attribute, program throw an exception.
```C# System.ArgumentException: 'Object of type 'System.String' cannot be converted to type 'System.Int32'.'```
As example i used System.Int32 as type.

Problem is at converting Id generated by GenerateId() method.

```class RedisObjectHandler.cs```
```C#
internal static string SetId(this object obj)
    {
      Type type = obj.GetType();
      DocumentAttribute customAttribute = Attribute.GetCustomAttribute((MemberInfo) type, typeof (DocumentAttribute)) as DocumentAttribute;
      PropertyInfo propertyInfo = ((IEnumerable<PropertyInfo>) type.GetProperties()).FirstOrDefault<PropertyInfo>((Func<PropertyInfo, bool>) (x => Attribute.GetCustomAttribute((MemberInfo) x, typeof (RedisIdFieldAttribute)) != null));
      string str = customAttribute != null ? customAttribute.IdGenerationStrategy.GenerateId() : throw new MissingMemberException("Missing Document Attribute decoration");
      if (propertyInfo != (PropertyInfo) null)
      {
        Type propertyType = propertyInfo.PropertyType;
        if (!((IEnumerable<Type>) new Type[3]
        {
          typeof (string),
          typeof (Guid),
          typeof (Ulid)
        }).Contains<Type>(propertyType) && !propertyType.IsValueType)
          throw new InvalidOperationException("Software Defined Ids on objects must either be a string, ULID, Guid, or some other value type.");
        object obj1 = propertyInfo.GetValue(obj);
        if (!RedisObjectHandler.TypeDefaultCache.ContainsKey(propertyType))
          RedisObjectHandler.TypeDefaultCache.Add(propertyType, Activator.CreateInstance(propertyType));
        if (obj1?.ToString() != RedisObjectHandler.TypeDefaultCache[propertyType]?.ToString())
          str = propertyInfo.GetValue(obj).ToString();
        else if (propertyType == typeof (Guid))
          propertyInfo.SetValue(obj, (object) Guid.Parse(str));
        else if (propertyType == typeof (Ulid))
          propertyInfo.SetValue(obj, (object) Ulid.Parse(str));
        else
         propertyInfo.SetValue(obj, (object) str);
      }
      return customAttribute.Prefixes == null || string.IsNullOrEmpty(((IEnumerable<string>) customAttribute.Prefixes).FirstOrDefault<string>()) ? type.FullName + ":" + str : ((IEnumerable<string>) customAttribute.Prefixes).First<string>() + ":" + str;
    }
```

FIXED by using ```propertyInfo.SetValue(obj, Convert.ChangeType(str,info.PropertyType),null);```

```C#
internal static string SetId(this object obj)
    {
      Type type = obj.GetType();
      DocumentAttribute customAttribute = Attribute.GetCustomAttribute((MemberInfo) type, typeof (DocumentAttribute)) as DocumentAttribute;
      PropertyInfo propertyInfo = ((IEnumerable<PropertyInfo>) type.GetProperties()).FirstOrDefault<PropertyInfo>((Func<PropertyInfo, bool>) (x => Attribute.GetCustomAttribute((MemberInfo) x, typeof (RedisIdFieldAttribute)) != null));
      string str = customAttribute != null ? customAttribute.IdGenerationStrategy.GenerateId() : throw new MissingMemberException("Missing Document Attribute decoration");
      if (propertyInfo != (PropertyInfo) null)
      {
        Type propertyType = propertyInfo.PropertyType;
        if (!((IEnumerable<Type>) new Type[3]
        {
          typeof (string),
          typeof (Guid),
          typeof (Ulid)
        }).Contains<Type>(propertyType) && !propertyType.IsValueType)
          throw new InvalidOperationException("Software Defined Ids on objects must either be a string, ULID, Guid, or some other value type.");
        object obj1 = propertyInfo.GetValue(obj);
        if (!RedisObjectHandler.TypeDefaultCache.ContainsKey(propertyType))
          RedisObjectHandler.TypeDefaultCache.Add(propertyType, Activator.CreateInstance(propertyType));
        if (obj1?.ToString() != RedisObjectHandler.TypeDefaultCache[propertyType]?.ToString())
          str = propertyInfo.GetValue(obj).ToString();
        else if (propertyType == typeof (Guid))
          propertyInfo.SetValue(obj, (object) Guid.Parse(str));
        else if (propertyType == typeof (Ulid))
          propertyInfo.SetValue(obj, (object) Ulid.Parse(str));
        else
          propertyInfo.SetValue(obj, Convert.ChangeType(str,info.PropertyType),null);
      }
      return customAttribute.Prefixes == null || string.IsNullOrEmpty(((IEnumerable<string>) customAttribute.Prefixes).FirstOrDefault<string>()) ? type.FullName + ":" + str : ((IEnumerable<string>) customAttribute.Prefixes).First<string>() + ":" + str;
    }
```

**Comment by Zlejsi** [comment]
@slorello89 - here is TestRepo with error
https://github.com/Zlejsi/TestRedisOM.git

**Comment by zulander1** [comment]
@Zlejsi , the try changing the Id from int to String! GenerateId is a string and your Id is int..

**Comment by slorello89** [comment]
@zulander1 put their finger on it, the IdGenerationStratagey API is an advanced use-case but it's intended to provide a string id to use in the key.

**Comment by Zlejsi** [comment]
Ok. I understand that it is advanced use case, but as i wrote before, you provide a way to set RedisIdField not only as string, but as different types also. I only paste my way to fix the issue to use field with integer value in it.

**Comment by zulander1** [comment]
@slorello89 maybe we should do a generic IIdGenerationStrategy<T>  and return as T, this should resolve this issue

**Comment by slorello89** [comment]
Should be fixed by #302
